server/eventHandlers/imOnlineEvents.ts

```diff
@@ -0,0 +1,102 @@
+import { IEventHandler, CWEvent, IChainEventData, SubstrateTypes } from '@commonwealth/chain-events';
+import { sequelize } from '../database';
+
+
+const uptimePercent = (noOfTrues: number, noOfFalse: number, currentEventType: number) => {
+  /*
+    This formula is used to calculate the uptime percentage of the validators based on previous sessions' uptime.
+    upTimePercentage = ((No. isOnline True(s) + Current Execution Mode [1 for AllGood, 0 for SomeOffline]) / 
+                        (No. isOnline True(s) + No. isOnline False(s) - 1 + 1)) * 100
+  */
+  let upTimePercentage = (((noOfTrues + currentEventType) / (noOfTrues + noOfFalse - 1 + 1)) * 100)
+    .toFixed(2);
+
+  return upTimePercentage;
+};
+
+export default class extends IEventHandler {
+  constructor(
+    private readonly _models
+  ) {
+    super();
+  }
+
+  /**
+    Event handler to store ImOnline information of validators details in DB.
+  */
+  public async handle(event: CWEvent < IChainEventData >, dbEvent) {
+    // 1) if other event type ignore and do nothing.
+    if (event.data.kind !== SubstrateTypes.EventKind.AllGood
+      && event.data.kind !== SubstrateTypes.EventKind.SomeOffline) {
+      return dbEvent;
+    }
+    const imOnlineEventData = event.data;
+    
+    // 2) Get relevant data from DB for processing.
+    /*
+      This query will return the last created record for validators in 'HistoricalValidatorStatistic' table and return the data for each validator with there onlineCount and offlineCount counts iff any.
+      since all new and active validators records has been created by new-session event handler, it'll return the the last created records of them.
+    */
+    const rawQuery = `
+      SELECT *
+      FROM( 
+        SELECT * ,ROW_NUMBER() OVER( PARTITION BY partitionTable.stash ORDER BY created_at DESC ) 
+        FROM public."HistoricalValidatorStatistic" as partitionTable
+        JOIN( 
+          SELECT stash, SUM(case when "isOnline" then 1 else 0 end) as "onlineCount", SUM(case when "isOnline"  then 0 else 1 end) as "offlineCount" 
+          FROM public."HistoricalValidatorStatistic" as groupTable GROUP by groupTable.stash
+          ) joinTable
+        ON joinTable.stash = partitionTable.stash
+        WHERE partitionTable.stash IN ('${imOnlineEventData.validators.join("','")}')
+        )  as validatorQuery
+      WHERE  validatorQuery.row_number = 1
+  `;
+    const [validators, metadata] = await sequelize.query(rawQuery);
+    let validatorsList = JSON.parse(JSON.stringify(validators));
+
+    // 3) Modify uptime for validators.
+    switch (imOnlineEventData.kind) {
```

Comment:
repeated code in switch. 
Move following lines outside switch, since they are the same in both switch cases.:
```
validator.isOnline = false
          validator.block = event.blockNumber.toString();
          validator.eventType = imOnlineEventData.kind;
          validator.created_at = new Date().toISOString();
          validator.updated_at = new Date().toISOString();
          delete validator.id;
          delete validator.onlineCount;
          delete validator.offlineCount;
          delete validator.row_number;
```